virtio-devices/src/device.rs

```diff
@@ -305,6 +312,11 @@ impl VirtioCommon {
         // requires the addresses held by the descriptors to be translated.
         self.avail_features &= !(1 << VIRTIO_F_RING_INDIRECT_DESC);
     }
+
+    #[cfg(all(feature = "mshv", feature = "sev_snp"))]
+    pub fn set_vm(&mut self, vm: Arc<dyn hypervisor::Vm>) {
+        self.vm = Some(vm);
```

Comment:
I'm really not sure about this - seems inappropriate to add a dependency between the virtio-devices onto the the hypervisor crate (and leak the "sev_snp" implementation details). It would be better the `impl AccessPlatform` on a new struct inside the vmm crate - where you can the initialise it with access to the VM fd, etc. This would mean you would not need to modify any virtio device code themselves. The only place you would need to change then is in DeviceManager where it instantiates the devices to pass your new `AccessPlatform` implementation.

Reply:
That's the reason I created this PR with minimal implementation on RNG device. The changes in this PR are already part of our internal fork and have a release. I do agree with your suggestions and it is cleaner.  But I really am not sure how can I implement AccessControl into vmm crate specially what new struct and what are the contents and references in the struct. Could you please elaborate a bit more?

Reply:
Of course - let me try and elaborate using some code snippets.

As you have spotted inside `virtio-devices` there is the following trait:

```rust
impl Translatable for GuestAddress {
    fn translate_gva(&self, access_platform: Option<&Arc<dyn AccessPlatform>>, len: usize) -> Self {
        GuestAddress(self.0.translate_gva(access_platform, len))
    }
    fn translate_gpa(&self, access_platform: Option<&Arc<dyn AccessPlatform>>, len: usize) -> Self {
        GuestAddress(self.0.translate_gpa(access_platform, len))
    }
}
```

This trait is implemented for both `GuestAddress` and `u64` - both these implementations then call into a struct implementing the `AccessPlatform` trait (the difference is that the `Translatable` trait takes a reference to the struct implementing `AccessPlatform` (i.e. `&Arc<dyn AccessPlatform>`)

The purpose of the `Translatable` trait is to make it easier use the `AccessPlatform` trait.

The original purpose of the `AccessPlatform` trait is to provide a way to translate from IO addresses to GPAs when the device is behind an IOMMU. As can be seen from the current (sole) implementer of the `AccessPlatform` trait - namely `AccessPlatformMapping` which is used to expose the the IOMMU in a clean way.

This is currently instantiated in the following code in `DeviceManager::add_virtio_device`:

```rust
        // Create the AccessPlatform trait from the implementation IommuMapping.
        // This will provide address translation for any virtio device sitting
        // behind a vIOMMU.
        let access_platform: Option<Arc<dyn AccessPlatform>> = if let Some(mapping) = iommu_mapping
        {
            Some(Arc::new(AccessPlatformMapping::new(
                pci_device_bdf.into(),
                mapping.clone(),
            )))
        } else {
            None
        };
```

Since you want to call into the `hypervisor::Vm` for all accesses from VIRTIO devices you can piggyback on the same infrastructure with something like:

```rust
impl AccessPlatform for SevSnpPageAccessProxy {
    fn translate_gpa(&self, base: u64, size: u64) -> std::result::Result<u64, std::io::Error> {
        Ok(base)
    }
    
    fn translate_gva(&self, base: u64, size: u64) -> std::result::Result<u64, std::io::Error> {
        self.vm.gain_page_access(base, size).map_err(...)?;
        Ok(base)
    }
}
```

If you do want to combine this with `virtio-iommu` you will want to chain into that.

Then you just need to modify `DeviceManager::add_virtio_device` to instantiate this new struct when you create the virtio device and pass it in. `DeviceManager::new` has `Arc<dyn hypervisor::Vm>` passed in so you will just need to modify `struct DeviceManager` to hold that so you can use it later.

With this approach the code will be a lot cleaner without more cross-crate dependencies.